text/0000-light-dom.md

```diff
@@ -0,0 +1,315 @@
+  [Styling is critical to web component reuse, but may prove difficult in practice](https://component.kitchen/blog/posts/styling-is-critical-to-web-component-reuse-but-may-prove-difficult-in-practice)
+
+- **Third party integrations**
+
+  Third party tools need to traverse the DOM, which breaks with Shadow DOM and the existing browser APIs (querySelector, ...). Note that the use of Light DOM fixes for Light DOM components, but not for native Shadow DOM ones if the page contains any: Analytics tools, Personalization platforms, Commerce tools like PriceSpider or Honey...
+
+- **Testing software**
+
+  Tools like Selenium, Cypress etc. face the same issues as third party tools when it comes to traversing the DOM.
+
+Introducing components rendering to the Light DOM opens new possibilities where applications can render most of their content in the Light DOM, but also keep individual widgets (e.g. `select` tag) and leaf components that render in the Shadow DOM. This approach combines the best of both worlds by solving the issues presented above and offering strong encapsulation when needed.
+![shadow spectrum](./shadow-spectrum.png?raw=true "Shadow Spectrum")
+
+### Prior Art
+
+Most of the libraries designed to support Shadow DOM also propose a Light DOM option, with a variety of Shadow DOM features (slots, scoped styles, etc.). It includes:
+
+- [StencilJS](https://stenciljs.com/docs/styling#shadow-dom-in-stencil)
+- [LitElement](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)
+- [MS Fast Element](https://fast.design/docs/fast-element/working-with-shadow-dom#shadow-dom-configuration)
+
+## Detailed design
+
+### `LightningElement.shadow`
+
+Toggle between light DOM and shadow DOM is done via a new `shadow` static property on the `LightningElement` constructor. This accepts a `boolean` value and is `true` by default. When set to `true`, the LightningElement creates a shadow root on the host element and renders the component content in the shadow root. When set to `false` the component renders its content directly in the host element light DOM.
+
+```js
+import { LightningElement } from "lwc";
+
+// Example of a shadow DOM component
+class ShadowDOMComponent extends LightningElement {}
+
+// Example of a light DOM component
+class LightDOMComponent extends LightningElement {
+  static shadow = false;
+}
+
+// Default value
+console.log(LightningElement.shadow); // true
+```
+
+The `shadow` property is looked up by the LWC engine when the component is instantiated to determine how it should render. Changing the value of the `shadow` static property after the instantiation doesn't influence whether components are rendered in the light DOM or in the shadow DOM.
```

Comment:
Does it influence components created after the `shadow` property's value is changed?

Should this value be made non-writable at runtime?

Reply:
Because of [how we're handling Babel loose mode](https://github.com/salesforce/lwc/blob/f0fb460fd3cacd71e6b702cdb5aa6343491216a1/packages/%40lwc/engine-core/src/framework/base-lightning-element.ts#L603-L615), I'm not sure it's possible to enforce this. I could be wrong though.

As-is though, if `shadow` is redefined after the `constructor()` is called, then the change would have no effect.

Reply:
Hm, so it looks like it's possible, but we have two options:

1. Disallow setting `shadow` a second time.
2. If any instance of `FooElement` runs its `constructor()`, disallow setting `shadow` again.

The downside of the first is that users may legitimately want to set `shadow` more than once (complex app logic) before calling the constructor. The downside of the second is that it's a bit odd than a single instance of a class would have an impact on the entire class.

Reply:
In the current implementation updating the `shadow` attribute after the component is created has no effect. The attribute is read during the construction phase and then cached in the `VM` for the lifetime of the component.

Making the static field non-writable doesn't provide much guaranty if the field is defined as a getter: 

```js
import { LightningElement } from 'lwc';

export default class FooElement extends LightingElement {
  static get shadow() {
    return Math.random() > 0.5;
  }
}
```

We can for example cache the value `shadow` value after the `FooElement` instantiation and reuse it for all the future `FooElement` instantiations. I am not personally in favor of making the first component instantiation special.